http-client/src/main/java/io/micronaut/http/client/netty/DefaultHttpClient.java

```diff
@@ -3660,4 +3081,373 @@ private static class CurrentEvent {
+                } else { // error flow
+                    try {
+                        promise.tryFailure(makeErrorFromRequestBody(msg.status(), response));
+                        response.onComplete();
+                    } catch (HttpClientResponseException t) {
+                        promise.tryFailure(t);
+                        response.onComplete();
+                    } catch (Exception t) {
+                        response.onComplete();
+                        promise.tryFailure(makeErrorBodyParseError(msg, httpStatus, t));
+                    }
+                }
+            } catch (HttpClientResponseException t) {
+                promise.tryFailure(t);
+            } catch (Exception t) {
+                makeNormalBodyParseError(msg, httpStatus, t, cause -> {
+                    if (!promise.tryFailure(cause) && log.isWarnEnabled()) {
+                        log.warn("Exception fired after handler completed: " + t.getMessage(), t);
+                    }
+                });
+            }
+        }
+
+        /**
+         * Create a {@link HttpClientResponseException} from a response with a failed HTTP status.
+         */
+        private HttpClientResponseException makeErrorFromRequestBody(HttpResponseStatus status, FullNettyClientHttpResponse<?> response) {
+            if (errorType != null && errorType != HttpClient.DEFAULT_ERROR_TYPE) {
+                return new HttpClientResponseException(
+                        status.reasonPhrase(),
+                        null,
+                        response,
+                        new HttpClientErrorDecoder() {
+                            @Override
+                            public Argument<?> getErrorType(MediaType mediaType) {
+                                return errorType;
+                            }
+                        }
+                );
+            } else {
+                return new HttpClientResponseException(status.reasonPhrase(), response);
+            }
+        }
+
+        /**
+         * Create a {@link HttpClientResponseException} if parsing of the HTTP error body failed.
+         */
+        private HttpClientResponseException makeErrorBodyParseError(FullHttpResponse fullResponse, HttpStatus httpStatus, Throwable t) {
+            FullNettyClientHttpResponse<Object> errorResponse = new FullNettyClientHttpResponse<>(
+                    fullResponse,
+                    httpStatus,
+                    mediaTypeCodecRegistry,
+                    byteBufferFactory,
+                    null,
+                    false
+            );
+            // this onComplete call disables further parsing by HttpClientResponseException
+            errorResponse.onComplete();
+            return new HttpClientResponseException(
+                    "Error decoding HTTP error response body: " + t.getMessage(),
+                    t,
+                    errorResponse,
+                    null
+            );
+        }
```

Comment:
There's no coverage of this new method (there may not have been any previously though) 😉